addons/stock_account/models/stock_move_line.py

```diff
@@ -47,6 +47,8 @@ def _create_correction_svl(self, move, diff):
         if move._is_in() and diff > 0 or move._is_out() and diff < 0:
             move.product_price_update_before_done(forced_qty=diff)
             stock_valuation_layers |= move._create_in_svl(forced_quantity=abs(diff))
+            if move.product_id.cost_method == 'average':
+                move.product_id._run_fifo_vacuum(move.company_id)
```

Comment:
- we now run the fifo vacuum at [each receipt](https://github.com/odoo/odoo/blob/13.0/addons/stock_account/models/stock_move.py#L276) and not on the scheduler/when opening the valuation report
- this method is called when unlock/edit a done move. incrementing a receipt and decrementing a delivery is considered as an incoming move, so you're right that it should also run in fifo to be completely correct
- the goal of the fifo vacuum is to "correct" the negative stock, i.e. you sent units you do not have so they were sent at an estimated price, then when a next receipt is validated the system tries to compute at which value the "negative delivery" should have been sent if it was sent "in the order" (when the more recent units are in stock).
- < v13, the remaining_qty/value fiels were updated in avco as if the valuation was fifo, what we forgot to do was to call the vacuum to actually correct the negative moves (the ones with a negative remaining_qty)
- \>= v13, the remaining_qty/remaining_value informations are now on the stock valuation layer and the method to compute the valuation/quantity valued is [the same across all valuation methods](https://github.com/odoo/odoo/blob/13.0/addons/stock_account/models/product.py#L101). this was because of this last change that we saw the issue of not doing the correction entries in avco.
- we thus need to call this method (this is a receipt + the receipt could create some corrections layers)

Reply:
you surely get the rationale: we consider the fifo order when fixing the negatice avco stock. we also consider the fifo order when applying a landed cost in avco [on a receipt where some units are already sent to a customer](https://github.com/odoo/odoo/blob/13.0/addons/stock_landed_costs/models/stock_landed_cost.py#L124) so it's consistent